sulong/projects/com.oracle.truffle.llvm.runtime/src/com/oracle/truffle/llvm/runtime/debug/debugexpr/nodes/DebugExprNodeFactory.java

```diff
@@ -0,0 +1,309 @@
+        checkError(pair, "cast");
+        if (pair.getType() == type) {
+            return pair;
+        }
+        LLVMExpressionNode node;
+        if (type.isUnsigned())
+            node = contextReference.get().getNodeFactory().createUnsignedCast(pair.getNode(), type.getLLVMRuntimeType());
+        else
+            node = contextReference.get().getNodeFactory().createSignedCast(pair.getNode(), type.getLLVMRuntimeType());
+        return DebugExpressionPair.create(node, type);
+    }
+
+    @SuppressWarnings("static-method")
+    public DebugExpressionPair createObjectMember(DebugExpressionPair receiver, String fieldName) {
+        Object baseMember = null;
+        String memberName = null;
+        if (receiver.getNode() instanceof DebugExprVarNode) {
+            baseMember = ((DebugExprVarNode) (receiver.getNode())).getMember();
+            memberName = ((DebugExprVarNode) (receiver.getNode())).getName();
+        } else if (receiver.getNode() instanceof DebugExprObjectMemberNode) {
+            baseMember = ((DebugExprObjectMemberNode) (receiver.getNode())).getMember();
+            memberName = ((DebugExprObjectMemberNode) (receiver.getNode())).getFieldName();
+        }
+        if (baseMember != null) {
+            DebugExprObjectMemberNode node = new DebugExprObjectMemberNode(fieldName, baseMember);
+            return DebugExpressionPair.create(node, node.getType());
+        }
+        throw DebugExprException.symbolNotFound(receiver.getNode(), memberName, null);
+    }
+
+    public DebugExpressionPair createFunctionCall(DebugExpressionPair functionPair, List<DebugExpressionPair> arguments) {
+        checkError(functionPair, "call(...)");
+        if (functionPair.getNode() instanceof DebugExprVarNode) {
+            DebugExprVarNode varNode = (DebugExprVarNode) functionPair.getNode();
+            LLVMExpressionNode node = varNode.createFunctionCall(arguments, globalScopes);
+            return DebugExpressionPair.create(node, varNode.getType());
+        }
+        throw DebugExprException.typeError(functionPair.getNode(), functionPair.getNode().toString());
+    }
+
+    @SuppressWarnings("static-method")
+    public DebugExpressionPair createArrayElement(DebugExpressionPair array, DebugExpressionPair index) {
+        Object baseMember = null;
+        DebugExprType baseType = null;
+        if (array.getNode() instanceof DebugExprVarNode) {
+            baseMember = ((DebugExprVarNode) (array.getNode())).getMember();
+            baseType = ((DebugExprVarNode) (array.getNode())).getType();
+        } else if (array.getNode() instanceof DebugExprObjectMemberNode) {
+            baseMember = ((DebugExprObjectMemberNode) (array.getNode())).getMember();
+            baseType = ((DebugExprObjectMemberNode) (array.getNode())).getType();
+        }
+        if (baseMember != null) {
+            DebugExprArrayElementNode node = new DebugExprArrayElementNode(baseMember, index.getNode(), baseType.getInnerType());
```

Comment:
I don't think the `DebugExprArrayElementNode` should get the `baseMember` as argument, it should rather have the `array.getNode()` as a child node. You shouldn't store runtime values in the AST. It's not really avoidable in the `DebugExprVarNode`, but we should at least avoid it where possible. In this case, the `DebugExprArrayElementNode` can call `execute` on the `array.getNode()` at runtime.

See also the comment in `DebugExprArrayElementNode`.